include/proteus/Frontend/Func.hpp

```diff
@@ -275,15 +276,979 @@ template <typename RetT, typename... ArgT> class Func final : public FuncBase {
+  return *this;
+}
+
+template <typename T>
+template <typename U>
+Var<T, std::enable_if_t<std::is_arithmetic_v<T>>> &
+Var<T, std::enable_if_t<std::is_arithmetic_v<T>>>::operator=(
+    const U &ConstValue) {
+  static_assert(std::is_arithmetic_v<U>,
+                "Can only assign arithmetic types to Var");
+
+  Type *LHSType = getValueType();
+
+  if (LHSType->isIntegerTy()) {
+    storeValue(ConstantInt::get(LHSType, ConstValue));
+  } else if (LHSType->isFloatingPointTy()) {
+    storeValue(ConstantFP::get(LHSType, ConstValue));
+  } else {
+    PROTEUS_FATAL_ERROR("Unsupported type");
+  }
+
+  return *this;
+}
+
+template <typename T>
+template <typename U>
+Var<std::common_type_t<T, U>>
+Var<T, std::enable_if_t<std::is_arithmetic_v<T>>>::operator+(
+    const Var<U> &Other) const {
+  return binOp(
+      *this, Other,
+      [](IRBuilderBase &B, Value *L, Value *R) { return B.CreateAdd(L, R); },
+      [](IRBuilderBase &B, Value *L, Value *R) { return B.CreateFAdd(L, R); });
+}
+
+template <typename T>
+template <typename U>
+Var<std::common_type_t<T, U>>
+Var<T, std::enable_if_t<std::is_arithmetic_v<T>>>::operator-(
+    const Var<U> &Other) const {
+  return binOp(
+      *this, Other,
+      [](IRBuilderBase &B, Value *L, Value *R) { return B.CreateSub(L, R); },
+      [](IRBuilderBase &B, Value *L, Value *R) { return B.CreateFSub(L, R); });
+}
+
+template <typename T>
+template <typename U>
+Var<std::common_type_t<T, U>>
+Var<T, std::enable_if_t<std::is_arithmetic_v<T>>>::operator*(
+    const Var<U> &Other) const {
+  return binOp(
+      *this, Other,
+      [](IRBuilderBase &B, Value *L, Value *R) { return B.CreateMul(L, R); },
+      [](IRBuilderBase &B, Value *L, Value *R) { return B.CreateFMul(L, R); });
+}
+
+template <typename T>
+template <typename U>
+Var<std::common_type_t<T, U>>
+Var<T, std::enable_if_t<std::is_arithmetic_v<T>>>::operator/(
+    const Var<U> &Other) const {
+  return binOp(
+      *this, Other,
+      [](IRBuilderBase &B, Value *L, Value *R) { return B.CreateSDiv(L, R); },
+      [](IRBuilderBase &B, Value *L, Value *R) { return B.CreateFDiv(L, R); });
+}
+
+template <typename T>
+template <typename U>
+Var<std::common_type_t<T, U>>
+Var<T, std::enable_if_t<std::is_arithmetic_v<T>>>::operator%(
+    const Var<U> &Other) const {
+  return binOp(
+      *this, Other,
+      [](IRBuilderBase &B, Value *L, Value *R) { return B.CreateSRem(L, R); },
+      [](IRBuilderBase &B, Value *L, Value *R) { return B.CreateFRem(L, R); });
+}
+
+// Arithmetic operators with ConstValue
+template <typename T>
+template <typename U, typename>
+Var<std::common_type_t<T, U>>
+Var<T, std::enable_if_t<std::is_arithmetic_v<T>>>::operator+(
+    const U &ConstValue) const {
+  static_assert(std::is_arithmetic_v<U>,
+                "Can only add arithmetic types to Var");
+  Var<U> Tmp = Fn.defVar<U>(ConstValue, "tmp.");
+  return (*this) + Tmp;
+}
+
+template <typename T>
+template <typename U, typename>
```

Comment:
I find the unnamed typename not very readable. Why not do SFINAE on the return type:
```cpp
std::enable_if_t<std::is_arithmetic_v<U>, Var<std::common_type_t<T, U>>> 
```
(same change to the in-class definition)

Applies also to other arithmetic operators.